runtime/Cpp/runtime/src/atn/PredictionContext.cpp

```diff
@@ -152,16 +152,14 @@ Ref<PredictionContext> PredictionContext::mergeSingletons(const Ref<SingletonPre
     return rootMerge;
   }
 
-  Ref<PredictionContext> parentA = a->parent;
-  Ref<PredictionContext> parentB = b->parent;
   if (a->returnState == b->returnState) { // a == b
-    Ref<PredictionContext> parent = merge(parentA, parentB, rootIsWildcard, mergeCache);
+    Ref<PredictionContext> parent = merge(a->parent, b->parent, rootIsWildcard, mergeCache);
 
     // If parent is same as existing a or b parent or reduced to a parent, return it.
-    if (parent == parentA) { // ax + bx = ax, if a=b
```

Comment:
This an the following if statement appears to be a bug with the initial C++ runtime. Its doing pointer comparison, not object equality comparison.

Reply:
This is not a bug. Look at the original Java code:

			// if parent is same as existing a or b parent or reduced to a parent, return it
			if ( parent == a.parent ) return a; // ax + bx = ax, if a=b
			if ( parent == b.parent ) return b; // ax + bx = bx, if a=b

That's a pointer comparison, not object equality. Object equality in Java is done via `parent.equals(a.parent)`. Here the real parent is required. Object equality is used where a set of properties is compared to other objects with that set of properties.

Reply:
Sorry my bad, I should have checked that against the original run-time.